src/pybind/mgr/ssh/module.py

```diff
@@ -409,25 +345,43 @@ def get_hosts(self):
         nodes = [orchestrator.InventoryNode(host_name, []) for host_name in self.inventory_cache]
         return orchestrator.TrivialReadCompletion(nodes)
 
-    def _get_device_inventory(self, host):
+    def _run_ceph_daemon(self, host, entity, command, args, stdin=None):
         """
-        Query storage devices on a remote node.
-
-        :return: list of InventoryDevice
+        Run ceph-daemon on the remote host with the given command + args
```

Comment:
I would really have loved to have a proper usage of remoto here by importing ceph-daemon as a Python libarary and then use remoto to transfer the Python code transparently for us. 

The current approach passing the file manually really looks ugly to me :wink: 

Maybe someting like

```python
import ceph_daemon
from remoto import connection
conn = connection.get('ssh').Connection(hostname)
remote_ceph_daemon = conn.import_module(ceph_daemon)
remote_ceph_daemon.main(...)
```

Some clear benefits are:

* Much less hacky
* No need to have ceph_daemon being a single python file
* No need to know where `ceph_daemon.py` is located, we just need to import it.

Reply:
We tried to make this work but kept running into problems that were difficult to debug.  (I forget if even the trivial case worked or not.)

In the end, I was worried about the complexity of the solution with no real benefit aside from cosmetics.  Also, note that we still want ceph-daemon (or at least a subset of it) as a single file so that you can wget/curl + exec it to bootstrap.

Reply:
I am neutral to weather remoto is used or not, but in the case we aren't the implementation needs to: 

- bubble up errors
- optionally allow to mute: errors, logging
- log commands being run
- log all errors
- deal with all bare excepts (with proper exception handling)
- allow granular process control: terminate on error, ignore on error, log/mute output
- remove all hardcoded executable paths
- if ssh is used, further customization of connections is needed (custom ssh conf, users other than root, etc...)

Those are the easy ones, there are probably more. It is tricky to get this right, and it shouldn't be treated as complete because the example workflow works.

I just realized that this has merged already. Will probably need tickets to address the missing pieces in the implementation